File: 56 Merge Intervals/Java/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> merged = new ArrayList<>();
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        int[] previous = intervals[0];
        for (int i = 1; i < intervals.length; i++) {
            int[] current = intervals[i];
            if (previous[1] >= current[0]) {
                previous[1] = Math.max(previous[1], current[1]);
            }
            else {
                merged.add(previous);
                previous = current;
            }
        }

        merged.add(previous);

        return merged.toArray(new int[merged.size()][]);
    }
}

File: 56 Merge Intervals/Python/Solution.py
from typing import List, Self


class Solution:
    def merge(self: Self, intervals: List[List[int]]) -> List[List[int]]:
        merged: List[List[int]] = []
        intervals.sort(key=lambda x: x[0])

        previous = intervals[0]
        for i in range(1, len(intervals)):
            current = intervals[i]

            if previous[1] >= current[0]:
                previous[1] = max(previous[1], current[1])
            else:
                merged.append(previous)
                previous = current

        merged.append(previous)
        
        return merged
